server/v2/stf/branch/writer_map.go

```diff
@@ -54,20 +54,40 @@ func (b WriterMap) ApplyStateChanges(stateChanges []store.StateChanges) error {
 }
 
 func (b WriterMap) GetStateChanges() ([]store.StateChanges, error) {
-	sc := make([]store.StateChanges, len(b.branchedWriterState))
-	for account, stateChange := range b.branchedWriterState {
-		kvChanges, err := stateChange.ChangeSets()
-		if err != nil {
-			return nil, err
-		}
+	var (
+		changes = make(map[string][]store.KVPair)
+		sc      []store.StateChanges
+	)
+	if err := b.recurseStateChanges(changes); err != nil {
+		return nil, err
+	}
+
+	for account, kvPairs := range changes {
 		sc = append(sc, store.StateChanges{
 			Actor:        []byte(account),
-			StateChanges: kvChanges,
+			StateChanges: kvPairs,
 		})
 	}
 	return sc, nil
 }
 
+func (b WriterMap) recurseStateChanges(changes map[string][]store.KVPair) error {
```

Comment:
not understanding really well this change

Reply:
I made some notes while this was in development, summarized them below.

![image](https://github.com/cosmos/cosmos-sdk/assets/1236532/228c4eb5-48de-4033-a2bd-4113850c8400)
![image](https://github.com/cosmos/cosmos-sdk/assets/1236532/1a62f824-63e3-478e-9d72-6c6e5d492981)


This call to `GetStateChanges()` in `InitChain` isn't returning the changes from InitGenesis only from DeliverBlock.  From the debug state above... the higher level state, `b.state.branchedWriterState` is i think the InitGenesis changes, while the state at `b.branchedWriterState` is the DeliverBlock state.  Is this as bug in branching or should GetStateChanges() be doing recursive iteration?

I went with recursive collection of state changes.

Reply:
i think branching is behaving as expected here.
it is just that we should be merging the "branch tree".

the changeset of branch_2 over branch_1 (which is a branch over state_0) is the diff between branch_1 and branch_2.

Or we add an explicit method, recursive apply changeset.

Reply:
or we:

first merge https://github.com/cosmos/cosmos-sdk/blob/kocu/simapp-v2/server/v2/appmanager/appmanager.go#L34 into https://github.com/cosmos/cosmos-sdk/blob/kocu/simapp-v2/server/v2/appmanager/appmanager.go#L43
and rename the second zeroState to `stateAfterGenesis`

then we merge `deliverBlockState` to `stateAfterGenesis`

^ this seems the most correct to me

Reply:
Based on the usage pattern so far, I think `core/store.GetStateChanges` should return changes from the leaf branch as well as all ancestors, since that's the more common behavior, and there should be a (an extension interface maybe?) of `GetBranchChanges()`, wdyt